lib/src/command/dependency_services.dart

```diff
@@ -407,8 +470,27 @@ class DependencyServicesApplyCommand extends PubCommand {
 class _PackageVersion {
   String name;
   Version? version;
+  String? gitRevision;
   VersionConstraint? constraint;
-  _PackageVersion(this.name, this.version, this.constraint);
+  _PackageVersion(this.name, String? versionOrHash, this.constraint)
+      : version = versionOrHash == null ? null : tryParseVersion(versionOrHash),
+        gitRevision =
+            versionOrHash == null ? null : tryParseHash(versionOrHash);
+}
+
+Version? tryParseVersion(String v) {
+  try {
+    return Version.parse(v);
+  } on FormatException {
+    return null;
+  }
+}
+
+String? tryParseHash(String v) {
+  if (RegExp(r'^[a-fA-F0-9]+$').hasMatch(v)) {
+    return v;
+  }
+  return null;
 }
```

Comment:
Do these need to be exported? can't they just be private?

Reply:
They might as well be private.